d3a_interface/matching_algorithms/pay_as_clear_matching_algorithm.py

```diff
@@ -83,13 +84,15 @@ def _smooth_discrete_point_curve(obj, limit, asc_order=True):
                 obj[i] = obj.get(i, 0) + obj.get(i + 1, 0)
         return obj
 
-    def _get_clearing_point(self, max_rate):
+    @staticmethod
+    def _get_clearing_point(max_rate: int, cumulative_bids: OrderedDict,
+                            cumulative_offers: OrderedDict) -> (int, float):
+        """Gets the energy rate and cumulative energy at the point of equilibrium."""
         for rate in range(1, max_rate + 1):
-            if self.state.cumulative_offers[rate] >= self.state.cumulative_bids[rate]:
-                if self.state.cumulative_bids[rate] == 0:
-                    return rate-1, self.state.cumulative_offers[rate-1]
-                else:
-                    return rate, self.state.cumulative_bids[rate]
+            if cumulative_offers[rate] >= cumulative_bids[rate]:
+                if cumulative_bids[rate] == 0:
+                    return rate-1, cumulative_offers[rate-1]
```

Comment:
Could you explain what this does? Why is it `rate - 1`?

Reply:
This needs familiarity with the 2 sided PAC discrete point curve algorithm to understand. Bids and offers are sorted according to their energy rate, each creating a series of lines in a 2-dimensional graph. This method traverses this graph by increments of 1 cents / kWh , and tries to find the intersection of the offer and bid lines (at the point when the offer becomes greater than the bid line, that's when we know that the lines intersect). This is a special case that, in reality, will appear only if bids and offers are both zero at one energy rate, meaning that the energy needs to be sold for zero rate. In this case we select the energy rate just before the last one, in order to not trade energy for free.